include/pybind11/embed.h

```diff
@@ -87,29 +89,112 @@ struct embedded_module {
     }
 };
 
+/// Python 2.x/3.x-compatible version of `PySys_SetArgv`
+inline void set_interpreter_argv(int argc, char** argv, bool add_current_dir_to_path) {
+    // Before it was special-cased in python 3.8, passing an empty or null argv
+    // caused a segfault, so we have to reimplement the special case ourselves.
+    char** safe_argv = argv;
+    std::unique_ptr<char*[]> argv_guard;
+    std::unique_ptr<char[]> argv_inner_guard;
+    if (nullptr == argv || argc <= 0) {
+        argv_guard = std::unique_ptr<char*[]>(safe_argv = new char*[1]);
+        argv_inner_guard = std::unique_ptr<char[]>(safe_argv[0] = new char[1]);
+        safe_argv[0][0] = '\0';
+        argc = 1;
+    }
+    size_t argv_size = static_cast<size_t>(argc);
+#if PY_MAJOR_VERSION >= 3
+    // SetArgv* on python 3 takes wchar_t, so we have to convert.
+    std::unique_ptr<wchar_t*[]> widened_argv(new wchar_t*[argv_size]);
+#  if PY_MINOR_VERSION >= 5
+    // Use of PyMem_RawFree here instead of PyMem_Free is as recommended by the python
+    // API docs: https://docs.python.org/3/c-api/sys.html#c.Py_DecodeLocale
+    struct pymem_rawfree_deleter {
+        void operator()(void* ptr) const {
+            PyMem_RawFree(ptr);
+        }
+    };
+    std::vector< std::unique_ptr<wchar_t[], pymem_rawfree_deleter> > widened_argv_entries;
+#  else
+    std::vector< std::unique_ptr<wchar_t[]> > widened_argv_entries;
```

Comment:
@bstaletic is there a way to avoid the `std::vector` I've added here and still get smart-pointer deletion?